diff --git a/ConfigureChecks.cmake b/ConfigureChecks.cmake
index e0826caf..734eba28 100644
--- a/ConfigureChecks.cmake
+++ b/ConfigureChecks.cmake
@@ -37,7 +37,7 @@ check_function_exists(gettimeofday HAVE_GETTIMEOFDAY)
 check_function_exists(localtime_r HAVE_LOCALTIME_R)
 check_function_exists(popen HAVE_POPEN)
 check_function_exists(mkstemp HAVE_MKSTEMP)
-check_function_exists(strtok_r HAVE_STRTOK_R)
+check_symbol_exists(strtok_r "string.h" HAVE_STRTOK_R)
 
 macro(CHECK_FOR_DIR include var)
   check_c_source_compiles(
